### PR TITLE
Add 2 patterns to fuse nodes into SoftmaxCrossEntropyLoss

### PR DESCRIPTION
**Pull request by xadupre**
Fixes infinite outputs when this operator is decomposed.

### CODE REVIEW

File: onnxscript/rewriter/onnx_fusion_rule_sets_test.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    ):
+        # import onnxruntime does not because of the subfolder named onnxruntime
+        import os
+        import sys
+
+        sys_path = sys.path
+        sys.path = [os.path.join(onnx.__file__, "..", "..")] + [
+            p for p in sys_path if "onnxruntime" in p
+        ]
+        import onnxruntime
```
Comment (justinchuby) [inline onnxscript/rewriter/onnx_fusion_rule_sets_test.py:58]: hmm

File: onnxscript/rewriter/pattern.py [hunk 2]
```diff
@@ -1036,7 +1038,10 @@ def _bind_value(self, pattern_value: ValuePattern, value: ir.Value) -> bool:
                 # TODO(rama): Use appropriate equality-check here: future extension possibility.
                 if match.bindings[pattern_value.name] == value:
                     return True
-                return self.fail(f"Variable {pattern_value.name} is bound to multiple values.")
+                return self.fail(
+                    f"Variable {pattern_value.name} is bound to multiple values "
```
Comment (justinchuby) [inline onnxscript/rewriter/pattern.py:1042]:
```suggestion
                    f"Variable {pattern_value.name} is bound to multiple values. "
```

File: onnxscript/rewriter/pattern.py [hunk 3]
```diff
@@ -1036,7 +1038,10 @@ def _bind_value(self, pattern_value: ValuePattern, value: ir.Value) -> bool:
...
+                    f"value is {value}, bound value is {match.bindings[pattern_value.name]}."
+                )
```
Comment (justinchuby) [inline onnxscript/rewriter/pattern.py:1044]:
```suggestion
                    f"value='{value}', bound_value='{match.bindings[pattern_value.name]}'"
                )
```

File: onnxscript/rewriter/onnx_fusion_rule_sets.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+from __future__ import annotations
+
+import onnx
+import onnx.numpy_helper
```
Comment (justinchuby) [inline onnxscript/rewriter/onnx_fusion_rule_sets.py:6]:
```suggestion
```

File: onnxscript/rewriter/onnx_fusion_rule_sets.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        neq3 = op.Not(op.Equal(indices, -100))
+        wh1 = op.Where(neq1, indices, 0)
+        uns = op.Unsqueeze(wh1, 1)
+        ge = op.GatherElements(
+            op.LogSoftmax(X, axis=1), op.Cast(uns, to=onnx.TensorProto.INT64), axis=1
+        )
+        wh2 = op.Where(neq2, op.Neg(op.Squeeze(ge, 1)), 0)
+        denominator = op.Cast(
+            op.ReduceSum(
+                op.Cast(neq3, to=onnx.TensorProto.INT64),
```
Comment (justinchuby) [inline onnxscript/rewriter/onnx_fusion_rule_sets.py:65]:
```suggestion
                op.Cast(neq3, to=ir.DataType.INT64),
```

I would avoid any protobuf reference in the rewrite rules.

File: onnxscript/rewriter/onnx_fusion_rule_sets.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+from __future__ import annotations
+
+import onnx
```
Comment (justinchuby) [inline onnxscript/rewriter/onnx_fusion_rule_sets.py:5]:
```suggestion
from onnxscript import ir
```

File: onnxscript/rewriter/onnx_fusion_rule_sets_test.py (hunk 1 above)
Comment (xadupre) [reply onnxscript/rewriter/onnx_fusion_rule_sets_test.py:58]: The right fix is to rename onnxruntime subfolder.

Comment (xadupre) [reply onnxscript/rewriter/onnx_fusion_rule_sets_test.py:58]: See #1786.

File: onnxscript/rewriter/pattern.py (hunk 3 above)
Comment (xadupre) [reply onnxscript/rewriter/pattern.py:1044]: done

**Comment by justinchuby** [comment]
The pattern should now become

{LogSoftmax -> NegativeLogLikelihoodLoss}

**Pull request closed without merging**